restic-exporter.py

```diff
@@ -10,149 +10,228 @@
-                client["hostname"],
-                client["username"],
-                client["snapshot_hash"]
-            ]
-            backup_timestamp.add_metric(common_label_values, client["timestamp"])
-            backup_files_total.add_metric(common_label_values, client["files_total"])
-            backup_size_total.add_metric(common_label_values, client["size_total"])
-            backup_snapshots_total.add_metric(common_label_values, client["snapshots_total"])
+        for client in self.metrics["clients"]:
+            backup_meta.add_metric(
+                [
+                    client["hostname"],
+                    client["username"],
+                    client["snapshot_hash"],
+                    client["snapshot_tag"],
+                ],
+                1,
+            )
+            backup_timestamp.add_metric([client["snapshot_hash"]], client["timestamp"])
+            backup_files_total.add_metric(
+                [client["snapshot_hash"]], client["files_total"]
+            )
+            backup_size_total.add_metric(
+                [client["snapshot_hash"]], client["size_total"]
+            )
+            backup_snapshots_total.add_metric(
+                [client["snapshot_hash"]], client["snapshots_total"]
+            )
+
+        scrape_duration_seconds.add_metric([], self.metrics["duration"])
 
         yield check_success
         yield snapshots_total
         yield backup_timestamp
+        yield backup_meta
         yield backup_files_total
         yield backup_size_total
         yield backup_snapshots_total
+        yield scrape_duration_seconds
 
-    def refresh(self, exit_on_error=False):
+    def refresh(self, any_error_fatal=False):
         try:
             self.metrics = self.get_metrics()
         except Exception:
-            logging.error("Unable to collect metrics from Restic. %s", traceback.format_exc(0).replace("\n", " "))
-            if exit_on_error:
+            logging.error(
+                "Unable to collect metrics from Restic. %s",
+                traceback.format_exc(0).replace("\n", " "),
+            )
+
+            # Shutdown exporter for any error
+            if any_error_fatal:
                 sys.exit(1)
 
     def get_metrics(self):
+        duration = time.time()
         all_snapshots = self.get_snapshots()
         latest_snapshots = self.get_snapshots(True)
         clients = []
         for snap in latest_snapshots:
-            stats = self.get_stats(snap['id'])
+            # Collect stats for each snap only if enabled
+            if self.disable_stats:
+                # return zero as "no-stats" value
+                stats = {
+                    "total_size": -1,
+                    "total_file_count": -1,
+                }
+            else:
+                stats = self.get_stats(snap["id"])
+
+            # use first element of tags if tags is present
+            if "tags" in snap:
+                tag = snap["tags"][0]
```

Comment:
Why not add all tags? Prometheus query language allows regular expressions.

Reply:
Cause what to do with tags next? We could make `k,v` tags, for example:

```python
__name__{tag1="value1", tag2="value2"}
```

but tags is a one `list[]`, so all tags will be:

```
__name__{tags="tag1,tag2"}
```

What to do with this tags next, with alert? 
What do we do with tags?

* put e-mail address to tags - to this e-mail address alerts will be pushed (the actual owner of the backup)
* put application name to tags - send alerts to chat, where actual owner catch hist application
* use alertmanager routing for finding the receiver by app name/team name/project